---
title: "Fstats_experimental.Rmd"
author: "Kiran Bajaj"
output: html_document
date: "2025-02-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#setwd("~/Desktop/MVP-H2F-HatcheryField")
```

### Packages

```{r}

library(dplyr)
library(bigsnpr)
library(bigstatsr)
library(robustbase)
library(ggplot2)
library(ggrepel)
library(hierfstat)
library(reshape2)
library(adegenet)
library(viridis)
library(pegas)
library(tidyverse)
library(colorspace)
library(OutFLANK)
library(CMplot)

```

### Load data

```{r}

thinned_snps <- readRDS(file.path("..", "..", "data", "parental_genetics_structure", "20250203_THINNEDSNPs_exp.rds"))

full_snps <- readRDS(file.path("..", "..", "data", "parental_genetics_structure", "20250203_FULLSNPs_exp.rds"))

muts_maf_filtered <- readRDS(file.path("..", "..", "data", "parental_genetics_structure", "20250203_FULLmuts_exp.rds"))

samp_full_subset <- readRDS(file.path("..", "..", "data", "parental_genetics_structure", "20250203_samp_full_subset_exp.rds"))

```

```{r}

# rename datasets for processing
full <- full_snps
gen <- thinned_snps
muts <- muts_maf_filtered

# check dimensions
dim(gen) # 154 individuals and 104958 SNPs
dim(muts) 

```

### Fst 

#### Data preparation

```{r}

# filter SNPs that are all heterozygous (all 1s)
keep_snps_fst <- apply(gen, 2, function(col) !all(col == 1))

# filter genotype matrix for these SNPs
GEN_filtered_fst <- gen[, keep_snps_fst]
dim(GEN_filtered_fst) # 1 column removed - now there are 104957 SNPs

# check column name to make sure it was a real column
col_heterozygous <- apply(gen, 2, function(col) all(col == 1))
colnames(gen)[col_heterozygous] #Affx-1246226248

```

```{r}

# now filter the muts again to just retain our SNPs that are not all heterozygous
muts_heterozygous_filtered <- muts[muts$Affx.ID %in% colnames(GEN_filtered_fst), ]
dim(muts_heterozygous_filtered) # goes down to 104957

```

``` {r}

# rename for processing
fst_data <- GEN_filtered_fst
head(fst_data)

# make sample IDs into a real column (instead of rownames)
sample_IDs <- rownames(fst_data)
fst_data$merge_ID_inds <- sample_IDs # "merge_ID_inds" is the column name in samp_full_subset

# reset the rownames to empty
rownames(fst_data) <- NULL

# make sure labels in fst data still line up with sample data
if(!identical((fst_data$merge_ID_inds),samp_full_subset$merge_ID_inds)){
    print("Error sample names don't line up")
    break
   } # passes

# store data to check labels 
fst_labels_check <- fst_data$merge_ID_inds
head(fst_labels_check)
samp_labels_check <- samp_full_subset$merge_ID_inds
head(samp_labels_check)

# check everything is all lined up # everything good
identical(fst_labels_check, samp_labels_check) # TRUE

```

```{r}

# add population labels to fst data
fst_data$ID_SiteDate <- samp_full_subset$ID_SiteDate

fst_data <- fst_data %>%
  mutate(ID_SiteDate = recode(ID_SiteDate,
                           "DEBY" = "S2-DEBY",
                           "FL-KingPlan" = "W3-FL",
                           "LA-SisLake" = "W2-LA",
                           "LOLA" = "S1-LOLA",
                           "ME-HogIs" = "W6-ME",
                           "NH-GrtBay" = "W5-NH",
                           "TX-CapBay" = "W1-TX",
                           "VA-DeepWatSh" = "W4-VA"))

# move sample and population columns to front 
n_cols <- ncol(fst_data)
fst_data <- fst_data[, c((n_cols-1):n_cols, 1:(n_cols-2))]

# sanity check: number of columns 
dim(fst_data) # number of SNPs (104957) + sample col + pop col = 104959

# check number of samples per population
population_counts <- table(fst_data$ID_SiteDate)
population_counts

```

```{r}

# remove sample ID column for Fst calculation
fst_data <- fst_data[, -1]

# convert population to factor
fst_data$ID_SiteDate <- as.factor(fst_data$ID_SiteDate)

# check dimensions one more time
dim(fst_data) # 104959 - pop col = 104958

```

#### Data product: Fst matrix 

```{r}

# run Fst calculation
#fst_results <- pairwise.WCfst(fst_data)
#fst_results[is.na(fst_results)] <- 0

# view results
#print(fst_results)

#saveRDS(fst_results, file.path("..", "..", "data", "pairwise_differences_analyses", "fst_results.rds"))

# load in fst results
fst_results <- readRDS(file.path("..", "..", "data", "pairwise_differences_analyses", "fst_results.rds"))


# convert the matrix to a long format, and edit column names
fst_plotting <- melt(fst_results, na.rm = TRUE)
colnames(fst_plotting) <- c("Var1", "Var2", "value")

# create a vector for the desired order of populations
population_order <- c("W1-TX", "W2-LA", "W3-FL", "S1-LOLA", "S2-DEBY", "W4-VA", "W5-NH", "W6-ME")

# reorder the Var1 and Var2 columns based on the defined order
fst_plotting$Var1 <- factor(fst_plotting$Var1, levels = population_order)
fst_plotting$Var2 <- factor(fst_plotting$Var2, levels = population_order)

# view data one more time
fst_plotting

# create heatmap using viridis color palette
fig_fst_heatmap <- ggplot(fst_plotting, aes(Var1, Var2, fill = value)) +
  geom_tile(color = "white") +
  #geom_text(aes(label = round(value, 3)), color = "black") +
  scale_fill_viridis_c(option = "D", na.value = "grey50") +  # use viridis color scale
  labs(x = "Population", y = "Population", fill = "Fst") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("fig_fst_heatmap.png", fig_fst_heatmap, width = 8, height = 6)

```

### Basic stats with hierfstat

#### Data preparation

```{r}

# rename for processing
basic_stats_data <- fst_data

# make the population column a factor
basic_stats_data$ID_SiteDate <- as.factor(basic_stats_data$ID_SiteDate)

# convert data frame to a genind object using df2genind
genind_obj<- df2genind(basic_stats_data[, -1],   # exclude population column
                        pop = basic_stats_data$ID_SiteDate, # provide population column here
                        ploidy = 2,             # specify ploidy (2 for diploids)
                        ncode = 1) 

```

```{r}

# convert to hierfstat object using genind2hierfstat
hierfstat_obj <- genind2hierfstat(genind_obj) 

# run basic stats
div <- basic.stats(hierfstat_obj) 

```

#### Observed and expected heterozygosity

```{r}

# extract values
Ho <- div$Ho # one value per SNP per population
Hs <- div$Hs # one value per SNP per population

# convert observed heterozygosity values to a data frame
Ho_df <- as.data.frame(Ho)
Ho_df$Locus <- rownames(Ho_df)

# convert expected heterozygosity values to a data frame
Hs_df <- as.data.frame(Hs)
Hs_df$Locus <- rownames(Hs_df)

# reshape data frames to long format
Ho_long_pop <- Ho_df %>%
  pivot_longer(cols = -Locus, names_to = "Population", values_to = "Ho") 
Hs_long_pop <- Hs_df %>%
  pivot_longer(cols = -Locus, names_to = "Population", values_to = "Hs")

# merge observed and expected data by populations and loci
pop_data <- merge(Ho_long_pop, Hs_long_pop, by = c("Locus", "Population"))

# combine into long format for plotting
pop_long <- pop_data %>%
  pivot_longer(cols = c("Ho", "Hs"), names_to = "Type", values_to = "Value")

# plot values by population
ggplot(pop_long, aes(x = factor(Population, levels = population_order), y = Value, fill = Type)) +
  geom_boxplot() +
  labs(title = "Observed and Expected Heterozygosity by Population",
       x = "Population",
       y = "Heterozygosity") +
  scale_fill_manual(values = c("Ho" = "blue", "Hs" = "orange")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 

```

```{r}

# calculate average observed and expected heterozygosity per population
avg_het_per_pop <- pop_data %>%
  group_by(Population) %>%
  summarise(
    avg_observed_ho = mean(Ho, na.rm = TRUE),
    avg_expected_hs = mean(Hs, na.rm = TRUE)
  )

print(avg_het_per_pop)

```

```{r}

# reshape average heterozygosity data for plotting
avg_het_long <- avg_het_per_pop %>%
  pivot_longer(cols = c(avg_observed_ho, avg_expected_hs), names_to = "Type", values_to = "Value")

# plot by population
ggplot(avg_het_long, aes(x = factor(Population, levels = population_order), y = Value, group = Type, color = Type)) +
  geom_line(size = 1.2) +  # lines for averages
  geom_point(size = 3) +   # points for each average
  labs(title = "Average Ho and Hs by Population",
       x = "Population",
       y = "Average Heterozygosity") +
  scale_color_manual(values = c("avg_observed_ho" = "blue", "avg_expected_hs" = "orange")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # angle labels for readability

```

#### Allelic richness

Total number of unique alleles present in a population. Results are close to whole numbers because of small adjustments from rarefaction. 

```{r}

# calculate allelic richness
ar_results <- allelic.richness(hierfstat_obj, min.n=34, diploid = TRUE) # min.n = number in smallest population * 2

# extract results
ar_results <- ar_results$Ar # one value per SNP per population

# make column for locus
ar_results <- rownames_to_column(ar_results, var = "Locus")
ar_results <- ar_results[, c("Locus", colnames(ar_results)[-1])] # move to front

# format data for plotting
ar_long <- melt(ar_results, id.vars = "Locus", variable.name = "Population", value.name = "Allelic_Richness")

# calculate means to put on graph
means <- aggregate(Allelic_Richness ~ Population, data = ar_long, mean)

```

```{r}

# plot allelic richness results
ggplot(ar_long, aes(x = factor(Population, levels = population_order), y = Allelic_Richness)) +
  geom_boxplot(fill = "lightblue") +
  labs(title = "Allelic Richness by Population", x = "Population", y = "Allelic Richness") +
  geom_point(data = means, aes(x = Population, y = Allelic_Richness, shape = "Mean"), 
             size = 2, color = "black") +  # plot means in addition to medians
  scale_shape_manual(name = "Legend", values = c(16), labels = c("Mean alleleic richness")) + 
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

```{r}

# calculate average ar per population
avg_ar_per_pop <- ar_long %>%
  group_by(Population) %>%
  summarise(
    avg_allelic_richness = mean(Allelic_Richness, na.rm = TRUE),
  )

print(avg_ar_per_pop)

```

#### Data product: AR, Ho, Hs

This will get used in a linear model that includes temperature, salinity, AR, and Ho. It is saved to the "parent_effects_H2F" folder under "data."

```{r}

parent_genetic_effects <- merge(avg_het_per_pop, avg_ar_per_pop, by = "Population") %>%
  mutate(Population = factor(Population, levels = population_order)) %>%
  arrange(Population)

write_csv(parent_genetic_effects, file.path("..", "..", "data", "parent_effects_H2F", "parent_genetic_effects.csv"))

```

### Fst outliers with OutFLANK

#### Data preparation

Same steps as before but with full SNP set. 

```{r}

# prepare the full SNP set
full <- full_snps

dim(full)

```

```{r}
# filter SNPs that are all heterozygous (all 1s)
keep_snps_fst <- apply(full, 2, function(col) !all(col == 1))

# filter genotype matrix for these SNPs
FULL_filtered_fst <- full[, keep_snps_fst]
dim(FULL_filtered_fst) # 1 column removed - now there are 154608 SNPs

# check column name to make sure it was a real column
col_heterozygous <- apply(full, 2, function(col) all(col == 1))
colnames(full)[col_heterozygous] #Affx-1246226248

# now filter the muts again to just retain our SNPs that are not all heterozygous
muts_outflank <- muts[muts$Affx.ID %in% colnames(FULL_filtered_fst), ]
dim(muts_outflank) # goes down to 154608

```

```{r}

outflank_data <- FULL_filtered_fst

# make sample IDs into a real column (instead of rownames)
sample_IDs <- rownames(outflank_data)
outflank_data$merge_ID_inds <- sample_IDs # "merge_ID_inds" is the column name in samp_full_subset

# reset the rownames to empty
rownames(outflank_data) <- NULL

# make sure labels in fst data still line up with sample data
if(!identical((outflank_data$merge_ID_inds),samp_full_subset$merge_ID_inds)){
    print("Error sample names don't line up")
    break
   } # passes

# store data to check labels 
fst_labels_check <- outflank_data$merge_ID_inds
head(fst_labels_check)
samp_labels_check <- samp_full_subset$merge_ID_inds
head(samp_labels_check)

# check everything is all lined up # everything good
identical(fst_labels_check, samp_labels_check) # TRUE

```

```{r}

# add population labels to outflank data
outflank_data$ID_SiteDate <- samp_full_subset$ID_SiteDate

outflank_data <- outflank_data %>%
  mutate(ID_SiteDate = recode(ID_SiteDate,
                           "DEBY" = "S2-DEBY",
                           "FL-KingPlan" = "W3-FL",
                           "LA-SisLake" = "W2-LA",
                           "LOLA" = "S1-LOLA",
                           "ME-HogIs" = "W6-ME",
                           "NH-GrtBay" = "W5-NH",
                           "TX-CapBay" = "W1-TX",
                           "VA-DeepWatSh" = "W4-VA"))

# move sample and population columns to front 
n_cols <- ncol(outflank_data)
outflank_data <- outflank_data[, c((n_cols-1):n_cols, 1:(n_cols-2))]

# sanity check: number of columns 
dim(outflank_data) # number of SNPs (154608) + sample col + pop col = 154610

# check number of samples per population
population_counts <- table(outflank_data$ID_SiteDate)
population_counts

```

```{r}

# remove sample ID column for Fst calculation
outflank_data <- outflank_data[, -1]

# convert population to factor
outflank_data$ID_SiteDate <- as.factor(outflank_data$ID_SiteDate)

# check dimensions one more time
dim(outflank_data) # 154610 - pop col = 154609

```

```{r}

# OutFLANK components: snp matrix, locus names, population names
# documentation: https://github.com/whitlock/OutFLANK/blob/master/OutFLANK%20readme.pdf

# create snp matrix
SNPmat <- as.matrix(outflank_data[,-1])  # snp matrix - exclude the population column
SNPmat <- apply(SNPmat, 2, as.numeric)  # convert all columns to numeric
SNPmat[is.na(SNPmat)] <- 9  # replace NA with 9 for missing data - there shouldn't be any 9s

# create locus names
locus_names <- colnames(outflank_data)[-1]  # exclude first column (population)

# create pop names
pop_names <- rep(outflank_data$ID_SiteDate, each = nrow(SNPmat) / nrow(outflank_data)) # replicate based on number of rows
pop_counts_outflank <- table(pop_names) # create a table to check with pop counts from before

pop_counts_outflank # these two should match
population_counts # these two should match

# make input data for OutFLANK
FstDataFrame <- MakeDiploidFSTMat(SNPmat, locus_names, pop_names)

```

#### Data checks

```{r}

# data check: heterozygosity vs. Fst
plot(FstDataFrame$He, FstDataFrame$FST) 

# data check: Fst vs. FstNoCorr
plot(FstDataFrame$FST, FstDataFrame$FSTNoCorr) # uncorrected Fst is always larger

```

#### Run OutFLANK

```{r}

# first, run OutFLANK on thinned SNPs
out_trim <- OutFLANK(FstDataFrame[FstDataFrame$LocusName %in% colnames(thinned_snps), ], # using pruned SNPs
                     LeftTrimFraction = 0.05, # defaults
                     RightTrimFraction = 0.05, 
                     Hmin = 0.1, 
                     NumberOfSamples = 8, # number of populations
                     qthreshold = 0.05)

# view results
str(out_trim)
head(out_trim$results)

# check the fit
OutFLANKResultsPlotter(out_trim, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         FALSE, RightZoomFraction = 0.05, titletext = NULL) # looks fine

# zoom in on right tail
OutFLANKResultsPlotter(out_trim , withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         TRUE, RightZoomFraction = 0.15, titletext = NULL) # looks fine

# check p value histogram - should look flat ish
hist(out_trim$results$pvaluesRightTail)

```

#### Calculate p-values

```{r}

# then, calulate p-values on the full set
P1 <- pOutlierFinderChiSqNoCorr(FstDataFrame, Fstbar = out_trim$FSTNoCorrbar, 
                                   dfInferred = out_trim$dfInferred, qthreshold = 0.05, Hmin=0.05)

# view data
head(P1)
tail(P1)

# look for outliers - highlighted in blue
my_out <- P1$OutlierFlag==TRUE
plot(P1$He, P1$FST, pch=19, col=rgb(0,0,0,0.1))
points(P1$He[my_out], P1$FST[my_out], col="blue")

# check the p-value histogram for the full set of data
# if there are outliers, it should look flat with an inflation near 0
hist(P1$pvaluesRightTail)

```

#### Data product: OutFLANK results, SNP info

```{r}

# separating by chromosome

a = sprintf("%08d", muts_outflank$Position)
muts_outflank$chrom_position <- paste(muts_outflank$Chromosome, a, sep = ".")

# add chromosome positions to dataframe
P1$Affx.ID <- P1$LocusName
joined_P1 <- left_join(P1, muts_outflank, by = c("Affx.ID" = "Affx.ID")) # add order check here

write_csv(joined_P1, file.path("..", "..", "data", "putative_selection", "outflank_results.csv"))

```
