---
title: "Fstats_experimental"
date: "2025-02-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Packages

```{r}

library(dplyr)
library(bigsnpr)
library(robustbase)
library(ggplot2)
library(ggrepel)
library(hierfstat)
library(reshape2)
library(adegenet)
library(viridis)
library(pegas)
library(tidyverse)
library(CMplot)
library(colorspace)
library(here)

```

### Load data

These three datasets are generated in "src/parental_genetics/genotypes_experimental.R"

```{r}

# thinned imputed SNP set
thin_snps <- readRDS(here("data", "parental_genetics_structure", "20250203_SNPs_exp.rds"))

# mutation data frame filtered for MAF
muts_maf_filtered <- readRDS(here("data", "parental_genetics_structure", "20250203_FULLmuts_exp.rds"))

# sample info
samp_full_subset <- readRDS(here("data", "parental_genetics_structure", "20250203_samp_full_subset_exp.rds"))

```

```{r}

# rename datasets for processing
snps <- thin_snps
muts <- muts_maf_filtered

# check dimensions
dim(snps) # 154 individuals and 43667 SNPs
dim(muts) 

snps
```

### Pairwise Fst

#### Data preparation

Filter SNPs that are all heterozygous. 

```{r}

# filter SNPs that are all heterozygous (all 1s)
keep_snps_fst <- apply(snps, 2, function(col) !all(col == 1))

# filter genotype matrix for these SNPs
GEN_filtered_fst <- gen[, keep_snps_fst]
dim(GEN_filtered_fst) # 1 column removed - now there are 43666 SNPs

# check column name to make sure it was a real column
col_heterozygous <- apply(gen, 2, function(col) all(col == 1))
colnames(gen)[col_heterozygous] #Affx-1246226248

```

```{r}

# now filter the muts again to just retain our SNPs that are not all heterozygous
muts_heterozygous_filtered <- muts[muts$Affx.ID %in% colnames(GEN_filtered_fst), ]
dim(muts_heterozygous_filtered) # goes down to 43666

```

### Format data for Fst

``` {r}

# rename for processing
fst_data <- GEN_filtered_fst

# make sample IDs into a real column (instead of rownames)
sample_IDs <- rownames(fst_data)
fst_data$merge_ID_inds <- sample_IDs # "merge_ID_inds" is the column name in samp_full_subset

# reset the rownames to empty
rownames(fst_data) <- NULL

# make sure labels in fst data still line up with sample data
if(!identical((fst_data$merge_ID_inds),samp_full_subset$merge_ID_inds)){
    print("Error sample names don't line up")
    break
   } # passes

# store data to check labels 
fst_labels_check <- fst_data$merge_ID_inds
head(fst_labels_check)
samp_labels_check <- samp_full_subset$merge_ID_inds
head(samp_labels_check)

# check everything is all lined up
identical(fst_labels_check, samp_labels_check) # TRUE

```

```{r}

# add population labels to fst data
fst_data$ID_SiteDate <- samp_full_subset$ID_SiteDate

# move sample and population columns to front 
n_cols <- ncol(fst_data)
fst_data <- fst_data[, c((n_cols-1):n_cols, 1:(n_cols-2))]

# sanity check: number of columns 
dim(fst_data) # number of SNPs (43666) + sample col + pop col = 43668

# check number of samples per population
population_counts <- table(fst_data$ID_SiteDate)
population_counts

```

```{r}

# remove sample ID column for Fst calculation
fst_data <- fst_data[, -1]

# convert population to factor
fst_data$ID_SiteDate <- as.factor(fst_data$ID_SiteDate)

# check dimensions one more time
dim(fst_data) # 43668 - pop col = 43667

```

## Calculate and plot Fst 

```{r}

# this is the line to calculate Fst - its commented out because I ran it on OOD
# fst_results <- pairwise.WCfst(fst_data)

# load in pairwise Fst results generated on OOD
fst_results <- read_csv(here("resultsexperimental", "fst_results_saved.csv"))

# view results
print(fst_results)

# convert the matrix to a long format, and edit column names
fst_plotting <- melt(fst_results, na.rm = TRUE)
colnames(fst_plotting) <- c("Var1", "Var2", "value")

# create a vector for the desired order of populations
population_order <- c("DEBY", "LOLA", "TX-CapBay", "LA-SisLake", "FL-KingPlan", "VA-DeepWatSh", "NH-GrtBay", "ME-HogIs")

# reorder the Var1 and Var2 columns based on the defined order
fst_plotting$Var1 <- factor(fst_plotting$Var1, levels = population_order)
fst_plotting$Var2 <- factor(fst_plotting$Var2, levels = population_order)

# view data one more time
fst_plotting

# create heatmap using viridis color palette
fig_fst_heatmap <- ggplot(fst_plotting, aes(Var1, Var2, fill = value)) +
  geom_tile(color = "white") +
  geom_text(aes(label = round(value, 3)), color = "black") +
  scale_fill_viridis_c(option = "D", na.value = "grey50") +  # use viridis color scale
  labs(title = "Pairwise Fst Heatmap", x = "Population", y = "Population") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("fig_fst_heatmap.png", fig_fst_heatmap, width = 8, height = 6)
```

# Basic stats with hierfstat

## Data preparation

```{r}

# rename for processing
basic_stats_data <- fst_data

# make the population column a factor
basic_stats_data$ID_SiteDate <- as.factor(basic_stats_data$ID_SiteDate)

# convert data frame to a genind object using df2genind
genind_obj<- df2genind(basic_stats_data[, -1],   # exclude population column
                        pop = basic_stats_data$ID_SiteDate, # provide population column here
                        ploidy = 2,             # specify ploidy (2 for diploids)
                        ncode = 1) 

```

```{r}

# convert to hierfstat object using genind2hierfstat
hierfstat_obj <- genind2hierfstat(genind_obj) 

# run basic stats
div <- basic.stats(hierfstat_obj) 

```

## Observed and expected heterozygosity

```{r}

# extract values
Ho <- div$Ho
Hs <- div$Hs

# convert observed heterozygosity values to a data frame
Ho_df <- as.data.frame(Ho)
Ho_df$Locus <- rownames(Ho_df)

# convert expected heterozygosity values to a data frame
Hs_df <- as.data.frame(Hs)
Hs_df$Locus <- rownames(Hs_df)

# reshape data frames to long format
Ho_long_pop <- Ho_df %>%
  pivot_longer(cols = -Locus, names_to = "Population", values_to = "Ho") 
Hs_long_pop <- Hs_df %>%
  pivot_longer(cols = -Locus, names_to = "Population", values_to = "Hs")

# merge observed and expected data by populations and loci
pop_data <- merge(Ho_long_pop, Hs_long_pop, by = c("Locus", "Population"))

# combine into long format for plotting
pop_long <- pop_data %>%
  pivot_longer(cols = c("Ho", "Hs"), names_to = "Type", values_to = "Value")

# plot values by population
fig_heterozygosity_boxplot <- ggplot(pop_long, aes(x = factor(Population, levels = population_order), y = Value, fill = Type)) +
  geom_boxplot() +
  labs(title = "Observed and Expected Heterozygosity by Population",
       x = "Population",
       y = "Heterozygosity") +
  scale_fill_manual(values = c("Ho" = "blue", "Hs" = "orange")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 

ggsave("fig_heterozygosity_boxplot.png", fig_heterozygosity_boxplot, width = 8, height = 6)

```

```{r}

# calculate average observed and expected heterozygosity per population
avg_het_per_pop <- pop_data %>%
  group_by(Population) %>%
  summarise(
    Avg_Observed_Ho = mean(Ho, na.rm = TRUE),
    Avg_Expected_Hs = mean(Hs, na.rm = TRUE)
  )

print(avg_het_per_pop)
write_csv(avg_het_per_pop, here("resultsexperimental", "avg_het_per_pop.csv"))

```

```{r}

# reshape average heterozygosity data for plotting
avg_het_long <- avg_het_per_pop %>%
  pivot_longer(cols = c(Avg_Observed_Ho, Avg_Expected_Hs), names_to = "Type", values_to = "Value")

# plot by population
fig_heterozygosity_by_population <- ggplot(avg_het_long, aes(x = factor(Population, levels = population_order), y = Value, group = Type, color = Type)) +
  geom_line(size = 1.2) +  # lines for averages
  geom_point(size = 3) +   # points for each average
  labs(title = "Average Ho and Hs by Population",
       x = "Population",
       y = "Average Heterozygosity") +
  scale_color_manual(values = c("Avg_Observed_Ho" = "blue", "Avg_Expected_Hs" = "orange")) +  # 
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # angle labels for readability

ggsave("fig_heterozygosity_by_population.png", fig_heterozygosity_by_population, width = 8, height = 6)

```

## Allelic richness

Total number of unique alleles present in a population. Results are close to whole numbers because of small adjustments from rarefaction. 

```{r}

# calculate allelic richness
ar_results <- allelic.richness(hierfstat_obj, min.n=34, diploid = TRUE) # min.n = number in smallest population * 2

# extract results
ar_results <- ar_results$Ar

# make column for locus
ar_results <- rownames_to_column(ar_results, var = "Locus")
ar_results <- ar_results[, c("Locus", colnames(ar_results)[-1])] # move to front

# format data for plotting
ar_long <- melt(ar_results, id.vars = "Locus", variable.name = "Population", value.name = "Allelic_Richness")

# calculate means to put on graph
means <- aggregate(Allelic_Richness ~ Population, data = ar_long, mean)

```

```{r}

# plot allelic richness results
fig_allelic_richness <- ggplot(ar_long, aes(x = factor(Population, levels = population_order), y = Allelic_Richness)) +
  geom_boxplot(fill = "lightblue") +
  labs(title = "Allelic Richness by Population", x = "Population", y = "Allelic Richness") +
  geom_point(data = means, aes(x = Population, y = Allelic_Richness, shape = "Mean"), 
             size = 2, color = "black") +  # plot means in addition to medians
  scale_shape_manual(name = "Legend", values = c(16), labels = c("Mean alleleic richness")) + 
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("fig_allelic_richness.png", fig_allelic_richness, width = 8, height = 6)

```

```{r}

# calculate average ar per population
avg_ar_per_pop <- ar_long %>%
  group_by(Population) %>%
  summarise(
    Avg_Allelic_Richness = mean(Allelic_Richness, na.rm = TRUE),
  )

print(avg_ar_per_pop)
write_csv(avg_ar_per_pop, here("resultsexperimental", "avg_ar_per_pop.csv"))

```

## Fis

Fis is the deviation of observed heterozygosity from expected heterozygosity in a population under Hardy-Weinberg equilibrium.

This graph is not great. Need to figure out a better way to visualize it.

```{r}

# format Fis data for plotting
Fis <- div$Fis
Fis_df <- as.data.frame(Fis)

# check dimensions
print("Before removing NAs")
dim(Fis_df) # 43666

# remove NaNs
Fis_df <- na.omit(Fis_df) # what point is best to remove missing data???

print("After removing NAs")
dim(Fis_df) # 22035 - this went down a lot

# add column for locus
Fis_df$Affx.ID <- rownames(Fis_df)

# pivot long
Fis_long <- Fis_df %>%
  pivot_longer(cols = -Affx.ID, names_to = "Population", values_to = "Fis") 

# join with chromosome positions
Fis_long$Affx.ID <- gsub("\\.", "-", Fis_long$Affx.ID)
joined_Fis <- left_join(Fis_long, muts_heterozygous_filtered, by = c("Affx.ID" = "Affx.ID"))
joined_Fis <- joined_Fis %>%
  select(Affx.ID, Population, Fis, Position, Group, organism, cust_gene_id, cust_gene_name)

head(joined_Fis)

Fis_plot <- ggplot(data = joined_Fis, aes(x = Position, y = Fis, group = Population, color = Population)) +
  geom_point() +             
  labs(title = "Fis by Position and Population",
       x = "Position",
       y = "Fis") +
  theme_minimal() 

# ggsave("Fis_plot.pdf", Fis_plot, width = 40, height = 30)

```

# Fst outliers with OutFLANK

## Data preparation

```{r}

library(OutFLANK)

outflank_data <- basic_stats_data

# OutFLANK components: snp matrix, locus names, population names
# documentation: https://github.com/whitlock/OutFLANK/blob/master/OutFLANK%20readme.pdf

# create snp matrix
SNPmat <- as.matrix(outflank_data[,-1])  # snp matrix - exclude the population column
SNPmat <- apply(SNPmat, 2, as.numeric)  # convert all columns to numeric
SNPmat[is.na(SNPmat)] <- 9  # replace NA with 9 for missing data - there shouldn't be any 9s

# create locus names
locus_names <- colnames(outflank_data)[-1]  # exclude first column (population)

# create pop names
pop_names <- rep(outflank_data$ID_SiteDate, each = nrow(SNPmat) / nrow(outflank_data)) # replicate based on number of rows
pop_counts_outflank <- table(pop_names) # create a table to check with pop counts from before

pop_counts_outflank # these two should match
population_counts # these two should match

# make input data for OutFLANK
FstDataFrame <- MakeDiploidFSTMat(SNPmat, locus_names, pop_names)

```

## OutFLANK data checks

```{r}

# data check: heterozygosity vs. Fst
plot(FstDataFrame$He, FstDataFrame$FST) 

# data check: Fst vs. FstNoCorr
plot(FstDataFrame$FST, FstDataFrame$FSTNoCorr) # uncorrected Fst is always larger

# what should I exclude?

```
## Trimming for LD

```{r}

# following "Bonus" section of OutFLANKAnalysis.Rmd vignette

# install bigstatsr and bigsnpr
if (!("bigstatsr" %in% installed.packages())){install.packages("bigstatsr")}
library(bigstatsr)
if (!("bigsnpr" %in% installed.packages())){devtools::install_github("privefl/bigsnpr")}
library(bigsnpr)

# convert snp matrix to required format for bigsnpr
G <- add_code256(big_copy(as.matrix(SNPmat), type="raw"), code=bigsnpr:::CODE_012)

# data check - number of columns in SNP matrix = number of values in chromosome and position vectors
if (ncol(SNPmat) == length(muts_heterozygous_filtered$Chromosome) && ncol(SNPmat) == length(muts_heterozygous_filtered$Position)) {
  cat("The number of SNPs matches the lengths of chromosome and position vectors.\n")
} else {
  cat("Mismatch detected:\n")
  cat("Number of SNPs: ", ncol(SNPmat), "\n")
  cat("Length of chromosome vector: ", length(muts_heterozygous_filtered$Chromosome), "\n")
  cat("Length of position vector: ", length(muts_heterozygous_filtered$Position), "\n")
}

# data check - order of locus names = order in SNP matrix and chromosome info

# check if the lengths are the same
if (length(muts_heterozygous_filtered$Affx.ID) != length(colnames(SNPmat))) {
  cat("Mismatch detected: Number of locus names does not match the number of SNP names.\n")
} else {
  # check if the orders are the same
  if (all(muts_heterozygous_filtered$Affx.ID == colnames(SNPmat))) {
    cat("The order of locus names matches the order of SNP matrix column names.\n")
  } else {
    cat("The order of locus names does NOT match the order of SNP matrix column names.\n")
    
    # identify mismatches
    mismatches <- muts_heterozygous_filtered$Affx.ID != colnames(SNPmat)
    cat("Mismatched names:\n")
    for (i in which(mismatches)) {
      cat("Position:", i, "- Locus Name:", muts_heterozygous_filtered$Affx.ID[i], "- SNP Name:", colnames(SNPmat)[i], "\n")
    }
  }
}

# SNP trimming
newpc <-snp_autoSVD(G=G,infos.chr =muts_heterozygous_filtered$Chromosome,infos.pos = muts_heterozygous_filtered$Position)
which_pruned <- attr(newpc, which="subset") # indexes of remaining SNPS after pruning

length(which_pruned) # 30939 - this got rid of kind of a lot

```

## Run OutFLANK with pruned SNPs

```{r}

# run OutFLANK
out_trim <- OutFLANK(FstDataFrame[which_pruned,], # using pruned SNPs
                     LeftTrimFraction = 0.05, # defaults
                     RightTrimFraction = 0.05, 
                     Hmin = 0.1, 
                     NumberOfSamples = 8, # number of populations
                     qthreshold = 0.05)

# view results
str(out_trim)
head(out_trim$results)

# check the fit
OutFLANKResultsPlotter(out_trim, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         FALSE, RightZoomFraction = 0.05, titletext = NULL) # looks fine

# zoom in on right tail
OutFLANKResultsPlotter(out_trim , withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         TRUE, RightZoomFraction = 0.15, titletext = NULL) # looks fine

# check p value histogram - should look flat ish
hist(out_trim$results$pvaluesRightTail)

```

## Calculate p-values for all loci

```{r}

# calulate p-values
P1 <- pOutlierFinderChiSqNoCorr(FstDataFrame, Fstbar = out_trim$FSTNoCorrbar, 
                                   dfInferred = out_trim$dfInferred, qthreshold = 0.05, Hmin=0.05)

# view data
head(P1)
tail(P1)

# look for outliers - highlighted in blue
my_out <- P1$OutlierFlag==TRUE
plot(P1$He, P1$FST, pch=19, col=rgb(0,0,0,0.1))
points(P1$He[my_out], P1$FST[my_out], col="blue")

# check the p-value histogram for the full set of data
# if there are outliers, it should look flat with an inflation near 0
hist(P1$pvaluesRightTail)

```

### Manhattan plot

```{r}

# separating by chromosome

a = sprintf("%08d", muts_heterozygous_filtered$Position)
muts_heterozygous_filtered$chrom_position <- paste(muts_heterozygous_filtered$Chromosome, a, sep = ".")

# add chromosome positions to dataframe
P1$Affx.ID <- P1$LocusName
joined_P1 <- left_join(P1, muts_heterozygous_filtered, by = c("Affx.ID" = "Affx.ID")) # add order check here
write_csv(joined_P1, here("resultsexperimental", "fst_muts_joined.csv"))

# create manhattan plot

pdf("fig_manhattan_plot.pdf", width = 40, height = 15) 

plot(joined_P1$chrom_position[joined_P1$He>0.1], joined_P1$FST[joined_P1$He>0.1],
     xlab="Position", ylab="FST", col=rgb(0,0,0,0.2))
  points(joined_P1$chrom_position[my_out], joined_P1$FST[my_out], col="magenta", pch=20) 
  
dev.off()

```

```{r}

# format data for CMplot
manhattan_pval <- joined_P1 %>%
  filter(He > 0.1) %>%  # filter based on heterozygosity
  dplyr::select(Affx.ID, Chromosome, Position, pvalues) %>%  # select relevant columns
  mutate(
    Chromosome = as.numeric(Chromosome),  # make sure Chromosome is numeric
    Position = as.numeric(Position)
  )

chr_colors <- viridis(length(unique(manhattan_pval$Chromosome)))
signal_colors <- lighten(chr_colors, amount = 0.7)  # lighten colors

CMplot(manhattan_pval,
       type = "p",                  
       plot.type = "m",            
       LOG10 = TRUE,
       cex = 0.6,
       col = chr_colors,           
       threshold = 0.01,     
       amplify = TRUE,
       signal.cex = 1,
       signal.col = signal_colors,
       file = "jpg",              
       file.name = "fst_pval_manhattan_plot", 
       dpi = 300,                  
       file.output = TRUE,         
       verbose = TRUE,             
       width = 20,                 
       height = 8,                 
       chr.labels.angle = 45)    

```

```{r}

# format data for CMplot
manhattan_fst <- joined_P1 %>%
  filter(He > 0.1) %>%  # filter based on heterozygosity
  dplyr::select(Affx.ID, Chromosome, Position, FST) %>%  # select relevant columns
  mutate(
    Chromosome = as.numeric(Chromosome),  # make sure Chromosome is numeric
    Position = as.numeric(Position),
    FST = ifelse(FST < 0, 0.00001, FST)         # Assign 0 to negative Fst values
  )

chr_colors <- viridis(length(unique(manhattan_fst$Chromosome)))
signal_colors <- lighten(chr_colors, amount = 0.7)  # lighten colors

CMplot(manhattan_fst,
       type = "p",                  
       plot.type = "m",            
       LOG10 = FALSE,
       cex = 0.6,
       col = chr_colors,           
       threshold = 0.4,     
       amplify = TRUE,
       signal.cex = 1,
       signal.col = signal_colors,
       file = "jpg",              
       file.name = "fst_manhattan_plot", 
       dpi = 300,                  
       file.output = TRUE,         
       verbose = TRUE,             
       width = 20,                 
       height = 8,                 
       chr.labels.angle = 45)    

```


