---
title: "GO_experimental.Rmd"
author: "Kiran Bajaj"
output: html_document
date: "2025-02-19"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#setwd("~/Desktop/MVP-H2F-HatcheryField")
```

### Packages

```{r}

library(tidyverse)
library(dplyr)
library(readr)
library(grid)
library(colorspace)

```

### Load data

```{r}

# Fst outliers generated by OutFLANK (using a q value of 0.05)
outflank_output <- read_csv(file.path("..", "..", "data", "putative_selection", "outflank_results.csv"))

```

### Gene enrichment

#### Data preparation

```{r}

# filter for outliers and SNPs with available gene info to put into ShinyGO

outliers_only <- outflank_output %>% # starts at 154102 rows
  filter(OutlierFlag == TRUE) %>% # down to 12077 flagged as outliers
  arrange(qvalues) %>% # sort by q value
  dplyr::select(LocusName, FST, pvalues, Chromosome, Position, cust_id, cust_gene_id, cust_gene_name) %>%
  filter(!is.na(cust_gene_id)) %>% # down to 929 with gene info available
  filter(!is.na(cust_gene_name)) %>%
  rename(gene = cust_gene_id)

# generate list for ShinyGO
shinygo_genes <- paste(outliers_only$gene, collapse = " ")
shinygo_genes # print list to copy and paste into ShinyGO

length(unique(outliers_only$gene)) # this is 868 unique genes

```

```{r}

# i'm interested in getting more annotations for this genome since only 929 out of 12077 outliers were annotated in the SNP dataframe that we got from the sequencing facility. 

# install relevant packages
#BiocManager::install(c("rtracklayer", "GenomicRanges"))

library(rtracklayer)
library(GenomicRanges)
library(dplyr)

# same code as above but don't filter for annotated SNPs this time
snp_df <- outflank_output %>% # starts at 154102 rows
  filter(OutlierFlag == TRUE) %>% # down to 12077 flagged as outliers
  arrange(qvalues) %>% # sort by q value
  dplyr::select(LocusName, FST, pvalues, qvalues, Chromosome, Position, cust_id, cust_gene_id, cust_gene_name)

# download the gff3 annotation from Ensembl, save it somewhere, and import it here
# download link: https://metazoa.ensembl.org/Crassostrea_virginica_gca002022765v4/Info/Index

gff <- rtracklayer::import(file.path("..", "..", "data", "putative_selection", "Crassostrea_virginica_gca002022765v4.C_virginica_3.0.61.gff3"))

```

```{r}

# our SNP dataframe uses numbers 1-10 for chromosomes but the gff3 object has a different naming scheme - we need to make these match
snp_df$scaffold <- dplyr::case_when(
  snp_df$Chromosome == 1 ~ "CM008241.1",
  snp_df$Chromosome == 2 ~ "CM008242.1",
  snp_df$Chromosome == 3 ~ "CM008243.1",
  snp_df$Chromosome == 4 ~ "CM008244.1",
  snp_df$Chromosome == 5 ~ "CM008245.1",
  snp_df$Chromosome == 6 ~ "CM008246.1",
  snp_df$Chromosome == 7 ~ "CM008247.1",
  snp_df$Chromosome == 8 ~ "CM008248.1",
  snp_df$Chromosome == 9 ~ "CM008249.1",
  snp_df$Chromosome == 10 ~ "CM008250.1"
)

# turn info from our SNP data frame into a GenomicRanges object - stores genomic coordinates in a standardized way
snp_gr <- GRanges(
  seqnames = snp_df$scaffold,
  ranges = IRanges(start = snp_df$Position, end = snp_df$Position),
  snp_id = snp_df$LocusName
)

# filter to get just the gene entries in the gff3 file (the file includes everything like mRNA, exons, etc.)
gene_features <- gff[gff$type == "gene"]

# find overlaps between our SNPs and genes in gff3 file
hits <- findOverlaps(snp_gr, gene_features)

# create an annotation data frame
annotated <- data.frame(
  snp_id = snp_gr$snp_id[queryHits(hits)],
  gene_id = mcols(gene_features)$ID[subjectHits(hits)]
)

# merge annotation data frame back with original 
snp_annotated <- left_join(snp_df, annotated, by = c("LocusName" = "snp_id"))

print(sum(!is.na(snp_annotated$gene_id))) # 8733 SNPs of 12077 now annotated with genes IDs

```

```{r}

# we now have way more gene IDs filled in. the next step is to connect to Ensembl online to query for biological information about our new gene IDs.

library(biomaRt)

# connect to the Ensembl Metazoa databse
mart <- useEnsemblGenomes(
  biomart = "metazoa_mart", 
  dataset = "cvgca002022765v4_eg_gene", # this is the code for the genome we want
  host = "https://metazoa.ensembl.org"
)

# clean up the annotated SNP data frame 
snp_annotated$gene_id <- as.character(snp_annotated$gene_id)
# remove the "gene:" prefix from new gene_id values
snp_annotated$gene_id <- sub("^gene:", "", snp_annotated$gene_id)
# remove NAs as before
gene_ids <- unique(na.omit(snp_annotated$gene_id))

# query Ensembl Metazoa for the information we want. attributes = columns to grab, filters = how you want to search, values = our gene IDs, mart = database connection
annotations <- getBM(
  attributes = c("ensembl_gene_id", "description", "go_id", "name_1006"),
  filters = "ensembl_gene_id",
  values = gene_ids,
  mart = mart
)

# ensembl_gene_id: the official Ensembl gene ID
# description: functional description
# go_id: Gene Ontology ID
# name_1006: name/definition of the GO term

# join back with SNP data using gene_id = ensembl_gene_id
snp_annotated <- left_join(snp_annotated, annotations, by = c("gene_id" = "ensembl_gene_id"))

# this creates a longer than expected data frame because it's creating a new row for every GO term - so SNPs appear in the df more than once. I want to collapse this.

```

```{r}

# collapse annotations to have a data frame with one row per SNP (harder to work with the GO term column with this format, but shinyGO will do it for us)

snp_annotated_collapsed <- snp_annotated %>%
  group_by(LocusName, gene_id, description) %>%   # group by the SNP identifier + gene_id + description
  summarise(
    # collapse the GO IDs (we're not using these anyway)
    go_id = {
      ids <- unique(na.omit(go_id))
      if (length(ids)) paste(ids, collapse = "; ") else NA_character_
    },
    # collapse the GO descriptions (we're not using these anyway)
    go_term = {
      terms <- unique(na.omit(name_1006))
      if (length(terms)) paste(terms, collapse = "; ") else NA_character_
    },
    .groups = "drop"
  )

# original SNP count (outliers)
nrow(snp_df) # 12077
# unique SNP IDs 
length(unique(snp_df$LocusName)) # 12077
# SNP count with gene info
length(unique(snp_annotated$LocusName)) # 12077
# unique SNP-gene pairs
length(unique(paste(snp_annotated$LocusName, snp_annotated$gene_id))) #12187

# I think this discrepancy comes from the fact that some SNPs are in multiple genes. I think it makes sense to keep these entries. 

# check how many SNPs appear in multiple genes
snp_gene_counts <- snp_annotated_collapsed %>%
  filter(!is.na(gene_id)) %>%    
  group_by(LocusName) %>%
  summarise(num_genes = n_distinct(gene_id)) %>%
  filter(num_genes > 1) # keep only SNPs in multiple genes

dim(snp_gene_counts) # 110 makes sense because 12187 - 12077 = 110. 

snp_annotated <- left_join(
  snp_df,
  snp_annotated_collapsed, # optional: drop 'description' if you want
  by = "LocusName"
  )

dim(snp_annotated) # 12187

```

```{r}

# generate list for ShinyGO 

snp_annotated_shinygo <- snp_annotated %>%
  filter(!is.na(gene_id))    

shinygo_genes <- paste(unique(snp_annotated_shinygo$gene_id), collapse = " ")
shinygo_genes # print list to copy and paste into ShinyGO

length(unique(snp_annotated_shinygo$gene_id)) # this is 3254 unique genes

```

#### ShinyGO

- Go to the ShinyGO website (ShinyGO)[https://bioinformatics.sdstate.edu/go/]
- Paste the list of genes above into the box on the left
- Press "Change species", type in crassostrea virginica, and choose: "cvirginica.3.0_eg_gene"
- Keep all other settings as default values and press submit
- On the "Enrichment" tab, download the top pathways
- On the "Genes" tab, download the gene info table

### Plotting

- Enriched: One row per enriched pathways. Corresponding genes in a column together, separated by spaces.
- Genes: One row per gene. Chromosome positions and additional details.

```{r}

# load in outputs from ShinyGO

# this one is the enriched pathways and the genes that go with them
enriched <- read_csv(file.path("..", "..", "data", "putative_selection", "shinyGO_enriched.csv"))

# this one is all the genes that matched the database and info about them
genes <- read_csv(file.path("..", "..", "data", "putative_selection", "shinyGO_genes.csv"))

```

#### Data preparation

```{r}

# prepare data
manhattan_fst <- outflank_output %>%
  filter(He > 0.1) %>%  # filter based on heterozygosity - goes down to 151340 SNPs
  mutate(
    Chromosome = as.numeric(Chromosome),
    Position = as.numeric(Position),
    qvalues = as.numeric(qvalues),
    FST = ifelse(FST < 0, 0.00001, FST),  # assign small value to negative FST
    Outlier = ifelse(qvalues < 0.05, "outlier", "nonoutlier"), # outliers are designated with q values 
    Color = ifelse(
      Outlier == "nonoutlier",
      ifelse(Chromosome %% 2 == 1, "#4A5763", "#6A747C"),  # base color
      lighten(ifelse(Chromosome %% 2 == 1, "#4A5763", "#6A747C"), 0.5) # lighten for outliers
    )
  )
# compute chromosome lengths and cumulative positions for plotting
chromosome_lengths <- manhattan_fst %>%
  group_by(Chromosome) %>%
  summarise(chr_len = max(Position)) %>%
  mutate(tot = cumsum(lag(chr_len + 1e7, default = 0))) # add a buffer between chromosomes for readability

# merge cumulative positions back into the data
manhattan_fst <- left_join(manhattan_fst, chromosome_lengths, by = c("Chromosome" = "Chromosome")) %>%
  arrange(Chromosome, Position) %>%
  mutate(Cumulative_Position = Position + tot)

# create axis ticks for chromosome labels
chrom_ticks <- manhattan_fst %>%
  group_by(Chromosome) %>%
  summarize(center = (max(Cumulative_Position) + min(Cumulative_Position)) / 2)

```

These data need to be reshaped. The OutFLANK data product has loci in rows, and genes IDs, brief functional annotations, chromosomes, and chromosome positions in columns. The "enriched" csv has pathways in rows, and a list of genes separated by spaces in one column that need to be split. The "genes" csv has genes in rows, and positions and bunch of other stuff in columns. We should check these positions with the OutFLANK positions (they might come from the same place so should be the same). 

```{r}

# split up the genes column so there is one row per gene
enriched_split <- enriched %>%
  mutate(gene = strsplit(as.character(Genes), " ")) %>%
  unnest(gene) %>%
  dplyr::select(-URL, -Genes)

# rename gene column for merge
genes <- genes %>%
  rename(gene = Pasted)

# join enriched pathway genes with full gene info from the other dataframe
enriched_joined <- enriched_split %>%
  inner_join(genes, by = "gene")

manhattan_fst <- manhattan_fst %>%
  rename(gene = cust_gene_id)

# join enriched pathway genes with Outflank output data
enriched_fully_joined <- manhattan_fst %>%
  inner_join(enriched_joined, by = "gene") %>%
  select(LocusName, `Enrichment FDR`, nGenes, `Pathway Genes`, `Fold Enrichment`, Pathway) %>%
  mutate(
    # extract GO_id
    GO_id = str_extract(Pathway, "^GO:\\d+"),
    # assign gene_type based on GO term grouping
    gene_type = case_when(
      GO_id %in% c("GO:0003700", "GO:0006355") ~ "Transcription and gene regulation",
      GO_id %in% c(
        "GO:0055085", "GO:0022857", "GO:0006811", "GO:0034220",
        "GO:0016020", "GO:0016021", "GO:0004930", "GO:0007186"
      ) ~ "Membrane and ion transport",
      GO_id %in% c("GO:0007165", "GO:0004672", "GO:0006468", "GO:0016310", "GO:0016301") ~ 
        "Signal transduction and kinase activity",
      GO_id %in% c("GO:0000166", "GO:0005524", "GO:0005509") ~ "Binding functions",
      GO_id == "GO:0005634" ~ "Nucleus",
      GO_id == "GO:0008081" ~ "Enzymatic activity",
      TRUE ~ "Other"
    )
  )

# # old pathways
#   # join enriched pathway genes with Outflank output data
# enriched_fully_joined <- manhattan_fst %>%
#   dplyr::inner_join(enriched_joined, by = "gene") %>%
#   dplyr::select(LocusName, `Enrichment FDR`, nGenes, `Pathway Genes`, `Fold Enrichment`, Pathway) %>%
#   dplyr::mutate(
#     # create categories based on individual GOs
#     gene_type = case_when(
#       str_extract(Pathway, "^GO:\\d+") %in% c("GO:0004806", "GO:0008083") ~ 
#         "Lipid metabolism and signaling", 
#       str_extract(Pathway, "^GO:\\d+") %in% c("GO:0043565", "GO:0003700", "GO:0006355") ~ 
#         "Gene expression and transcription",
#       str_extract(Pathway, "^GO:\\d+") %in% c("GO:0034220", "GO:0055085", "GO:0098655") ~ 
#         "Membrane and ion transport",
#       str_extract(Pathway, "^GO:\\d+") %in% c("GO:0007601", "GO:0004930", "GO:0007186") ~ 
#         "GPCR signaling and sensory",
#       str_extract(Pathway, "^GO:\\d+") %in% c("GO:0016021", "GO:0016020", "GO:0031012") ~ 
#         "Cellular structure and organization",
#       TRUE ~ "Other"
#     )
#   )

```

#### Manhattan plot

```{r}

# this is all the data we need for the manhattan plot
merged_manhattan <- manhattan_fst %>%
  left_join(enriched_fully_joined, by = "LocusName") # merge

# calculate where to draw the outlier cutoff line
threshold_fst <- min(merged_manhattan$FST[merged_manhattan$qvalues < 0.05], na.rm = TRUE)

# plotting
ggplot_manhattan <- ggplot(merged_manhattan, aes(x = Cumulative_Position, y = FST)) +
  
  # first layer: default points (background)
  geom_point(aes(color = Color), size = 1, shape = 1, alpha = 0.7) +  # slight transparency, open circles
  
  # second layer: highlighted points (foreground)
  geom_point(data = merged_manhattan %>% filter(!is.na(gene_type) & Outlier == "outlier"), 
            # add colors for GO categories
              aes(color = case_when(
               gene_type == "Transcription and gene regulation" ~ "#3e4989", # color by gene type
               gene_type == "Membrane and ion transport" ~ "#440154", 
               gene_type == "Signal transduction and kinase activity" ~ "#fde725",  
               gene_type == "Binding functions" ~ "#31688e",
               gene_type == "Nucleus" ~ "#35b779", 
               gene_type == "Enzymatic activity" ~ "#7b3294"
             )), size = 2) +  # closed circles on top
  
  # # second layer: highlighted points (foreground)
  # geom_point(data = merged_manhattan %>% filter(!is.na(gene_type) & Outlier == "outlier"), 
  #           # add colors for GO categories
  #             aes(color = case_when(
  #              gene_type == "Membrane and ion transport" ~ "#3e4989", # color by gene type
  #              gene_type == "GPCR signaling and sensory" ~ "#440154", 
  #              gene_type == "Lipid metabolism and signaling" ~ "#fde725",  
  #              gene_type == "Gene expression and transcription" ~ "#31688e",
  #              gene_type == "Cellular structure and organization" ~ "#35b779", 
  #            )), size = 2) +  # closed circles on top

  scale_x_continuous(breaks = chrom_ticks$center, labels = chrom_ticks$Chromosome) +
  scale_color_identity(guide = "none") +  # remove color legend
  labs(x = "Chromosome", y = expression(F[ST])) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, color = "black"),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(), 
    panel.grid.minor.y = element_blank(),
    axis.title = element_text(size = 16),    
    axis.text = element_text(size = 14),      
    legend.title = element_text(size = 15),      
    legend.text = element_text(size = 13)
  ) +
  geom_hline(aes(yintercept = threshold_fst, linetype = "q < 0.05"), color = "black") +
  scale_linetype_manual(name = "Outlier threshold", values = c("q < 0.05" = "dashed"))

# save plot
ggsave(file.path("..", "..", "figures", "putative_selection", "fst_manhattan_plot.pdf"), ggplot_manhattan, width = 15, height = 6, dpi = 300)

ggplot_manhattan

```

#### Lollipop plot

```{r}

# assign colors for GO categories
color_mapping <- c(
  "Transcription and gene regulation" = "#3e4989",
  "Membrane and ion transport" = "#440154",
  "Signal transduction and kinase activity" = "#fde725",
  "Binding functions" = "#31688e",
  "Nucleus" = "#35b779",
  "Enzymatic activity" = "#7b3294"
)

# lollipop plot
lollipop_plot <- ggplot(enriched_fully_joined, aes(x = `Fold Enrichment`, 
                                  y = reorder(Pathway, `Fold Enrichment`), 
                                  color = gene_type, 
                                  size = nGenes)) +
  geom_segment(aes(x = 0, xend = `Fold Enrichment`,  # set lengths of the segments
                   y = Pathway, yend = Pathway), 
               linewidth = 0.8) +  # width of the segments
  geom_point() +  # circles
  scale_color_manual(values = color_mapping) +  # assign colors
  scale_size_continuous(range = c(2, 5)) +  # adjust circle size range
  # set limits of the x scale to adjust to the length of the segments
  scale_x_continuous(expand = c(0, 0), limits = c(0, max(enriched_fully_joined$`Fold Enrichment`) * 1.1)) +  
  theme_bw() +
  theme( 
    axis.text = element_text(color = "black", size = 14),  # make axis text black
    panel.grid.major.x = element_blank(), # remove grid lines
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(), 
    panel.grid.minor.y = element_blank(), 
    axis.title = element_text(size = 16),    
    legend.title = element_text(size = 14),      
    legend.text = element_text(size = 13)
  ) +
  labs(x = "Fold enrichment", y = "Pathway", size = "Number of genes", color = "Pathway type") +
  guides( # customize shapes and colors of symbols on legend with an override
    color = guide_legend(override.aes = list(linetype = "blank", size = 3)),  
    size = guide_legend(override.aes = list(linetype = "blank", shape = 21, fill = "slategrey",
                                            color = "slategrey"))  
  )

# save figure
ggsave(file.path("..", "..", "figures", "putative_selection", "GO_lollipop_plot.pdf"), 
       lollipop_plot, width = 14, height = 6, dpi = 300)

lollipop_plot

```