---
title: "GO_experimental.Rmd"
author: "Kiran Bajaj"
output: html_document
date: "2025-02-19"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#setwd("~/Desktop/MVP-H2F-HatcheryField")
```

### Packages

```{r}

library(tidyverse)
library(dplyr)
library(readr)
library(grid)
library(colorspace)

```

### Load data

```{r}

# Fst outliers generated by OutFLANK
outflank_output <- read_csv(file.path("..", "..", "data", "putative_selection", "outflank_results.csv"))

```

### Gene enrichment

#### Data preparation

```{r}

# filter for outliers and SNPs with available gene info to put into ShinyGO
outliers_only <- outflank_output %>% # starts at 154608 rows
  filter(OutlierFlag == TRUE) %>% # down to 11863 flagged as outliers
  dplyr::select(LocusName, FST, pvalues, Chromosome, Position, cust_id, cust_gene_id, cust_gene_name) %>%
  filter(!is.na(cust_gene_id)) %>% # down to 917 with gene info available
  filter(!is.na(cust_gene_name)) %>%
  rename(gene = cust_gene_id)

# generate list for ShinyGO
shinygo_genes <- paste(outliers_only$gene, collapse = " ")
shinygo_genes # print list to copy and paste into ShinyGO

length(unique(outliers_only$gene)) # this is 862 unique genes

```

#### ShinyGO

- Go to the ShinyGO website (ShinyGO)[https://bioinformatics.sdstate.edu/go/]
- Paste the list of genes above into the box on the left
- Press "Change species", type in crassostrea virginica, and choose: "cvirginica.3.0_eg_gene"
- Keep all other settings as default values and press submit
- On the "Enrichment" tab, download the top pathways
- On the "Genes" tab, download the gene info table

### Plotting

- Enriched: One row per enriched pathways. Corresponding genes in a column together, separated by spaces.
- Genes: One row per gene. Chromosome positions and additional details.

```{r}

# load in outputs from ShinyGO

# this one is the enriched pathways and the genes that go with them
enriched <- read_csv(file.path("..", "..", "data", "putative_selection", "shinyGO_enriched.csv"))

# this one is all the genes that matched the database and info about them
genes <- read_csv(file.path("..", "..", "data", "putative_selection", "shinyGO_genes.csv"))


```

#### Data preparation

```{r}

# prepare data
manhattan_fst <- outflank_output %>%
  filter(He > 0.1) %>%  # filter based on heterozygosity - goes down to 151340 SNPs
  mutate(
    Chromosome = as.numeric(Chromosome),
    Position = as.numeric(Position),
    qvalues = as.numeric(qvalues),
    FST = ifelse(FST < 0, 0.00001, FST),  # assign small value to negative FST
    Outlier = ifelse(qvalues < 0.05, "outlier", "nonoutlier"), # outliers are designated with q values 
    Color = ifelse(
      Outlier == "nonoutlier",
      ifelse(Chromosome %% 2 == 1, "#4A5763", "#6A747C"),  # base color
      lighten(ifelse(Chromosome %% 2 == 1, "#4A5763", "#6A747C"), 0.5) # lighten for outliers
    )
  )
# compute chromosome lengths and cumulative positions for plotting
chromosome_lengths <- manhattan_fst %>%
  group_by(Chromosome) %>%
  summarise(chr_len = max(Position)) %>%
  mutate(tot = cumsum(lag(chr_len + 1e7, default = 0))) # add a buffer between chromosomes for readability

# merge cumulative positions back into the data
manhattan_fst <- left_join(manhattan_fst, chromosome_lengths, by = c("Chromosome" = "Chromosome")) %>%
  arrange(Chromosome, Position) %>%
  mutate(Cumulative_Position = Position + tot)

# create axis ticks for chromosome labels
chrom_ticks <- manhattan_fst %>%
  group_by(Chromosome) %>%
  summarize(center = (max(Cumulative_Position) + min(Cumulative_Position)) / 2)

```

These data need to be reshaped. The OutFLANK data product has loci in rows, and genes IDs, brief functional annotations, chromosomes, and chromosome positions in columns. The "enriched" csv has pathways in rows, and a list of genes separated by spaces in one column that need to be split. The "genes" csv has genes in rows, and positions and bunch of other stuff in columns. We should check these positions with the OutFLANK positions (they might come from the same place so should be the same). 

```{r}

# split up the genes column so there is one row per gene
enriched_split <- enriched %>%
  mutate(gene = strsplit(as.character(Genes), " ")) %>%
  unnest(gene) %>%
  dplyr::select(-URL, -Genes)

# rename gene column for merge
genes <- genes %>%
  rename(gene = Pasted)

# join enriched pathway genes with full gene info from the other dataframe
enriched_joined <- enriched_split %>%
  inner_join(genes, by = "gene")

manhattan_fst <- manhattan_fst %>%
  rename(gene = cust_gene_id)

# join enriched pathway genes with Outflank output data
enriched_fully_joined <- manhattan_fst %>%
  dplyr::inner_join(enriched_joined, by = "gene") %>%
  dplyr::select(LocusName, `Enrichment FDR`, nGenes, `Pathway Genes`, `Fold Enrichment`, Pathway) %>%
  dplyr::mutate(
    # create categories based on individual GOs
    gene_type = case_when(
      str_extract(Pathway, "^GO:\\d+") %in% c("GO:0004806", "GO:0008083") ~ 
        "Lipid metabolism and signaling", 
      str_extract(Pathway, "^GO:\\d+") %in% c("GO:0043565", "GO:0003700", "GO:0006355") ~ 
        "Gene expression and transcription",
      str_extract(Pathway, "^GO:\\d+") %in% c("GO:0034220", "GO:0055085", "GO:0098655") ~ 
        "Membrane and ion transport",
      str_extract(Pathway, "^GO:\\d+") %in% c("GO:0007601", "GO:0004930", "GO:0007186") ~ 
        "GPCR signaling and sensory",
      str_extract(Pathway, "^GO:\\d+") %in% c("GO:0016021", "GO:0016020", "GO:0031012") ~ 
        "Cellular structure and organization",
      TRUE ~ "Other"
    )
  )
```

#### Manhattan plot

```{r}

# this is all the data we need for the manhattan plot
merged_manhattan <- manhattan_fst %>%
  left_join(enriched_fully_joined, by = "LocusName") # merge

# calculate where to draw the outlier cutoff line
threshold_fst <- min(merged_manhattan$FST[merged_manhattan$qvalues < 0.05], na.rm = TRUE)

# plotting
ggplot_manhattan <- ggplot(merged_manhattan, aes(x = Cumulative_Position, y = FST)) +
  
  # first layer: default points (background)
  geom_point(aes(color = Color), size = 1, shape = 1, alpha = 0.7) +  # slight transparency, open circles
  
  # second layer: highlighted points (foreground)
  geom_point(data = merged_manhattan %>% filter(!is.na(gene_type) & Outlier == "outlier"), 
            # add colors for GO categories
              aes(color = case_when(
               gene_type == "Membrane and ion transport" ~ "#3e4989", # color by gene type
               gene_type == "GPCR signaling and sensory" ~ "#440154", 
               gene_type == "Lipid metabolism and signaling" ~ "#fde725",  
               gene_type == "Gene expression and transcription" ~ "#31688e",
               gene_type == "Cellular structure and organization" ~ "#35b779", 
             )), size = 2) +  # closed circles on top

  scale_x_continuous(breaks = chrom_ticks$center, labels = chrom_ticks$Chromosome) +
  scale_color_identity(guide = "none") +  # remove color legend
  labs(x = "Chromosome", y = "FST") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, color = "black"),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(), 
    panel.grid.minor.y = element_blank(), 
  ) +
  geom_hline(aes(yintercept = threshold_fst, linetype = "q < 0.05"), color = "black") +
  scale_linetype_manual(name = "Outlier threshold", values = c("q < 0.05" = "dashed"))

# save plot
ggsave(file.path("..", "..", "figures", "putative_selection", "fst_manhattan_plot.jpg"), ggplot_manhattan, 
       width = 15, height = 6, dpi = 300)

ggplot_manhattan

```

#### Lollipop plot

```{r}

# assign colors for GO categories
color_mapping <- c(
  "Membrane and ion transport" = "#3e4989",
  "GPCR signaling and sensory" = "#440154",
  "Lipid metabolism and signaling" = "#fde725",
  "Gene expression and transcription" = "#31688e",
  "Cellular structure and organization" = "#35b779"  
)

# lollipop plot
lollipop_plot <- ggplot(enriched_fully_joined, aes(x = `Fold Enrichment`, 
                                  y = reorder(Pathway, `Fold Enrichment`), 
                                  color = gene_type, 
                                  size = nGenes)) +
  geom_segment(aes(x = 0, xend = `Fold Enrichment`,  # set lengths of the segments
                   y = Pathway, yend = Pathway), 
               linewidth = 0.8) +  # width of the segments
  geom_point() +  # circles
  scale_color_manual(values = color_mapping) +  # assign colors
  scale_size_continuous(range = c(2, 5)) +  # adjust circle size range
  # set limits of the x scale to adjust to the length of the segments
  scale_x_continuous(expand = c(0, 0), limits = c(0, max(enriched_fully_joined$`Fold Enrichment`) * 1.1)) +  
  theme_bw() +
  theme( 
    axis.text = element_text(color = "black"),  # make axis text black
    panel.grid.major.x = element_blank(), # remove grid lines
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(), 
    panel.grid.minor.y = element_blank(), 
  ) +
  labs(x = "Fold enrichment", y = "Pathway", size = "Number of genes", color = "Gene type") +
  guides( # customize shapes and colors of symbols on legend with an override
    color = guide_legend(override.aes = list(linetype = "blank", size = 3)),  
    size = guide_legend(override.aes = list(linetype = "blank", shape = 21, fill = "slategrey",
                                            color = "slategrey"))  
  )

# save figure
ggsave(file.path("..", "..", "figures", "putative_selection", "GO_lollipop_plot.jpg"), 
       lollipop_plot, width = 10, height = 6, dpi = 300)

lollipop_plot

```